Fix playback and capture multiple runs which occasionally fails with Input/output error

iuliana-prodan opened the pull request:
This pull request fixes #3809.

The first commit updates the correct order for DAI and DMA start/stop.
Before this update, on multiple start/stop runs for playback or record or combined,
we get an underrun/overflow.
That's because the DAI makes a DMA request, before the DMA channel is
enabled.

The second commit updates the start/stop operations for SAI.
Please see a more detailed explanation in the commit message.

After the first 3 commits, after multiple playback/capture runs (~700, ~1000)
everything seems ok. 
But, sometimes, when running a second/third time, the same 700 loop I get an "ipc timed out for 0x60050000 size 12".
Here's a log:
```
Playing raw data '16/j16.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Playing raw data '16/j16.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Recording WAVE 'test_record.pcm' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
[  582.814139] sof-audio-of 556e8000.dsp: error: ipc timed out for 0x60050000 size 12
[  582.821802] sof-audio-of 556e8000.dsp: info: preventing DSP entering D3 state to preserve context
[  582.830906] sof-audio-of 556e8000.dsp: invalid header size 0x2000200. FW oops is bogus
[  582.838989] sof-audio-of 556e8000.dsp: error: unexpected fault 0x00000000 trace 0x00000000
[  582.847530] sof-audio-of 556e8000.dsp: error: waking up any trace sleepers
[  582.854601] sof-audio-of 556e8000.dsp: error: trace IO error
[  582.861198] sof-audio-of 556e8000.dsp: ASoC: error at snd_soc_pcm_component_trigger on 556e8000.dsp: -110
[  582.871466]  Port0: ASoC: trigger FE cmd: 0 failed: -110
Playing raw data '16/j16.wav' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Recording WAVE 'test_record.pcm' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
aplay: pcm_write:2061: write error: Input/output error
(null)      1  TFAIL  :  ltpapicmd.c:188: Test 420 Playback failed
Recording WAVE 'test_record.pcm' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
Recording WAVE 'test_record.pcm' : Signed 16 bit Little Endian, Rate 48000 Hz, Stereo
```
By looking in the trace I saw that the stop cmd is not triggered in DSP
and in SAI we get two start operations one after another.
Therefore, in the 3rd commit, we check in SAI, on start, if a start was already
triggered, disable just the DMA request and the data channel.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -264,6 +308,10 @@ static inline int sai_set_config(struct dai *dai,
+	dai_update_bits(dai, REG_SAI_XCR1(REG_RX_DIR),
+			REG_SAI_CR1_RFW_MASK, SAI_FIFO_WORD_SIZE / 2);
 	dai_update_bits(dai, REG_SAI_XCR2(REG_RX_DIR), mask_cr2, val_cr2);
 	dai_update_bits(dai, REG_SAI_XCR4(REG_RX_DIR), mask_cr4, val_cr4);
```
Ok, so you move FIFO_WORD_SIZE setting from sai_start to sai_set_config? I can understand this change but you need to add it in a separate patch and explain why is it needed. It doesn't look to be related to the current change.

iuliana-prodan replied:
Yes, I've missed this change. I'll add a new commit.
There's no need to set the FIFO size each time, on start.

dbaluta replied:
@iuliana-prodan you can add a new commit. That's an obvious change and if you want to be merged faster we can send it in a separate PR. 
For the rest of the changes lets give Intel people some time to look at this. I will also want to take my time to make some suggestions by the end of the week.

iuliana-prodan replied:
@dbaluta I've added a new commit just for the SAI watermark.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -33,39 +33,62 @@ static void sai_start(struct dai *dai, int direction)
 
 	uint32_t xcsr = 0U;
```
@ranj063 @plbossart I wonder if you can share your experience with IPC here. In stress tests, we get an IPC timeouts for stream STOP command.

iuliana-prodan replied:
@dbaluta if you refer to "ipc timed out", yes, you're right. I don't know exactly the root cause. The 3rd commit is optional, but the additional checks don't hurt :)

dbaluta commented on src/audio/dai.c:
```diff
@@ -651,11 +651,11 @@ static int dai_comp_trigger_internal(struct comp_dev *dev, int cmd)
 
 		/* only start the DAI if we are not XRUN handling */
```
@iuliana-prodan this is also the order ALSA uses to enable DAI and DMA, so in principle I agree with this change. Lets also hear Intel folks opinion on this as I remember they did the reverse change.

iuliana-prodan replied:
The order is not my opinion, this is what is written in i.MX reference manual (one can check for "Suspend/resume a DMA channel with active hardware service requests").

aiChaoSONG commented:
SOFCI TEST

keyonjie commented:
Looks this fails on some BYT platforms according to CI?

aiChaoSONG commented:
The failures on BYT are all related to pipeline `PCM: PCM Deep Buffer [hw:0,1]`

iuliana-prodan commented:
The only thing that might impact the CI tests it's the order from dai (commit 1).
I saw this commit 5c8c804766 ("dai: change starting sequence"), but it says "The purpose of this change is to allow slave interfaces to prepare their FIFOs before DMA starts transferring to them.".
I believe there are multiple Intel's topologies that uses slave interfaces, not just sof-byt-nocodec.tplg (this is from the failed test, but I cannot find it in sof repo).

Therefore, can somebody explain why, in the first place, the order for DAI and DMA was changed?
Where is this sof-byt-nocodec.tplg? Is it just for testing?

Thanks!

iuliana-prodan commented:
I looked further into the error and the problems seems to be on [dw_dma_stop](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L389), which expects for the [fifo to be cleared](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L423).
Most probably this is not drained anymore because the peripheral was stopped before the DMA.
Why do you need to clear the fifo? We just stopped the operation, can we just reset the fifo?
Or, isn’t this fifo emptied by default? Sorry, I don’t know the hardware. 

In the imx case, this is not applicable.

If there is no other way to empty the fifo, maybe we can add the “#if CONFIG_DMA_SUSPEND_DRAIN” in dai.c and, for this case, we keep the old order for DAI and DMA.

lgirdwood commented:
> @aiChaoSONG @keyonjie @lgirdwood @ranj063 @plbossart
> 
> I looked further into the error and the problems seems to be on [dw_dma_stop](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L389), which expects for the [fifo to be cleared](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L423).
> Most probably this is not drained anymore because the peripheral was stopped before the DMA.
> Why do you need to clear the fifo? 

The DW DMA needs the FIFO clear so it's in a empty state prior to next usage. i.e. no stale data is sent out.

We just stopped the operation, can we just reset the fifo?
> Or, isn’t this fifo emptied by default? Sorry, I don’t know the hardware.
> 
> In the imx case, this is not applicable.

ok, we need to differentiate then.

> 
> If there is no other way to empty the fifo, maybe we can add the “#if CONFIG_DMA_SUSPEND_DRAIN” in dai.c and, for this case, we keep the old order for DAI and DMA.

The best place for this is in the DMA driver and not any core logic. Btw, I dont see any reference to draining in dai.c, can you tell me where you would put the ifdef ?

iuliana-prodan commented:
> > If there is no other way to empty the fifo, maybe we can add the “#if CONFIG_DMA_SUSPEND_DRAIN” in dai.c and, for this case, we keep the old order for DAI and DMA.
> 
> The best place for this is in the DMA driver and not any core logic. Btw, I dont see any reference to draining in dai.c, can you tell me where you would put the ifdef ?

@lgirdwood There is already a CONFIG_DMA_SUSPEND_DRAIN in [dw/dma.c](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L415). 
I was thinking to add the same config (CONFIG_DMA_SUSPEND_DRAIN) in dai.c, in start and stop operations.
If the CONFIG_DMA_SUSPEND_DRAIN is set, we keep the order as before this change, otherwise we use the new order (added by commit 1 from this pull request).

dbaluta commented:
@lgirdwood Linux kernel ALSA order of DAI/DMA start is as @iuliana-prodan fixed in her patch. Any idea, why the ALSA with a non-dsp setup works but the DSP has problems?

lgirdwood commented:
> @lgirdwood Linux kernel ALSA order of DAI/DMA start is as @iuliana-prodan fixed in her patch. Any idea, why the ALSA with a non-dsp setup works but the DSP has problems?

Any update, could it be Pulseaudio searching for working configs on each PCM ?

lgirdwood commented:
> > If there is no other way to empty the fifo, maybe we can add the “#if CONFIG_DMA_SUSPEND_DRAIN” in dai.c and, for this case, we keep the old order for DAI and DMA.
> 
> The best place for this is in the DMA driver and not any core logic. Btw, I dont see any reference to draining in dai.c, can you tell me where you would put the ifdef ?

@iuliana-prodan So looking at the code again this is high level ordering and IIRC we also have a similar configuration in ASoC.

I would agree now and say here that we cant easily do in dma driver, it's a higher level framework CONFIG_ Kconfig setting. i.e. IMX should enable this in the Kconfig, I'm gathering patches for v1.7-final so we should add this fix.

dbaluta commented on src/audio/dai.c:
```diff
@@ -695,8 +695,21 @@ static int dai_comp_trigger_internal(struct comp_dev *dev, int cmd)
+		ret = dma_stop(dd->chan);
+#endif
 		break;
 	case COMP_TRIGGER_PAUSE:
```
@iuliana-prodan shouldn't we do the same thing at start? I mean the order of dai/dma start/stop should be in reverse one to each other.

iuliana-prodan replied:
At first, I thought the same thing.
But I've looked at the explanation for DMA_SUSPEND_DRAIN, from [Kconfig](https://github.com/thesofproject/sof/blob/master/Kconfig#L70) and this seems to happen only on stop.
Also, in the Intel's DMA driver this is only checked on [dw_dma_stop()](https://github.com/thesofproject/sof/blob/master/src/drivers/dw/dma.c#L389).

dbaluta replied:
Ok, I see. So, I'm fine with patches 1 and 2. I need to better review 3) and I think it is fine. I have my doubts about 4.

dbaluta commented on src/drivers/imx/sai.c:
```diff
@@ -74,32 +110,39 @@ static void sai_stop(struct dai *dai, int direction)
+			poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_PLAYBACK),
+						REG_SAI_CSR_TERE, 0, 100);
+		}
 	}
```
At least this part can be rewritten like this:

```
	/* Disable transmitter/receiver */
	if (direction == DAI_DIR_CAPTURE) {
		dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_CAPTURE), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_CAPTURE),
					REG_SAI_CSR_TERE, 0, 100);
	}
	
	
	/* Check if the opposite direction is also disabled */
	xcsr = dai_read(dai, REG_SAI_XCSR(!direction));
	if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(!direction), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(!direction),
						REG_SAI_CSR_TERE, 0, 100);
	}
```

iuliana-prodan replied:
It can be written like that only if this
```
if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(!direction), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(!direction),
						     REG_SAI_CSR_TERE, 0, 100);
}
```
changes to this:

```
if (!(xcsr & REG_SAI_CSR_FRDE)) {
		dai_update_bits(dai, REG_SAI_XCSR(DAI_DIR_PLAYBACK), REG_SAI_CSR_TERE, 0);
		poll_for_register_delay(dai_base(dai) + REG_SAI_XCSR(DAI_DIR_PLAYBACK),
						     REG_SAI_CSR_TERE, 0, 100);
}
```

iuliana-prodan replied:
As per our discussion, I've kept the original code for a clear/better understanding.

iuliana-prodan commented:
@dbaluta I've added a new version.
As per our discussion, I've dropped the 4th commit from my previous version (the extra checks for `ipc timed out` issue).
This needs more debugging and it will be handled in a separate pull request.

I've also separated the "sai start/stop operations" in 3 patches since some changes are not linked.

lgirdwood commented:
@zrombel this PR should have 0 impact on Intel systems but I do see it failing all codec PR tests ? Is this a false positive ?

iuliana-prodan commented:
@lgirdwood @dbaluta 
"All checks have passed" - I think is good to be merged :)

dbaluta commented:
Done. Iulia now we should backport these changes on 1.5 and 1.6 :D

iuliana-prodan commented:
> Done. Iulia now we should backport these changes on 1.5 and 1.6 :D

Yes, I will backport these patches.
Thanks!
